### ISSUE TITLE
IAM is blank when contains " ’ " character.

### ISSUE THREAD

**Issue by chanhpham9291** [Appboy/appboy-ios-sdk#201]
:white_check_mark: I have checked [Braze's documentation](https://www.braze.com/documentation/iOS/) and am still experiencing the issue.

# Report

## Describe your environment.

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 12.1
 SDK Version             | 3.20.1
 Integration Method      | cocoapods
 Xcode Version           | Xcode 10.1
 Repro rate              | all the time (100%)
 
## What did you do?

- Upgrade BrazeSDK from 3.7.1 to 3.20.1 via cocoapods.
- All IAMs will be blank if they contain this character: `’`
- Didn't see any error logs.

## What did you expect to happen?

The IAM is displayed normally in the device.

## What happened instead?

The IAM is blank in the device. Remove `’` or change it to `'` will fix the issue.

## Steps to reproduce

- Upgrade BrazeSDK from 3.7.1 to 3.20.1 via cocoapods.
- Trigger an IAM that has `’`.

## Code Snippet

` <p>Be the first to know about new content, events, updates, and more. Don’t miss out!</p>`

## Are you doing any feature customizations that may relate to the issue? Can you share the code snippet?

IAMs are displayed correctly in the old version (3.7.1).

**Comment by FarAndHeight** [comment]
+1

**Comment by chshapiro** [comment]
Hey @chanhpham9291 and @FarAndHeight,

We're unable to repro this issue. Are you integrating through Cocoapods?

Do you do any customization for in-app messages?

Thanks!

**Comment by FarAndHeight** [comment]
Hello @chshapiro we were able to reproduce this issue too.  We've just integrated 3.20.1 into our project as well and all our in app messages (that have a `’`, as @chanhpham9291 pointed out - we just tried to send one with and without the character and there is definitely a difference) display as blank screens.  

One thing to note, when we tried typing that character directly into the Braze dashboard in Chome - it didn't affect anything - maybe Chrome replaces it with a `'` or something - but when copying text from another text editor (we used Sublime to edit the html) - the  `’` character definitely caused an issue.

**Comment by chshapiro** [comment]
Hey @FarAndHeight,

I still can't repro the issue when copying from a text editor. Are you doing any customization for in-app messages?

**Comment by FarAndHeight** [comment]
@chshapiro Do you mean in the app or in the Braze dashboard?

**Comment by FarAndHeight** [comment]
@chshapiro In the Braze console/dashboard we do create custom in-app messages using HTML/CSS/JS + SVGs.

**Comment by chshapiro** [comment]
@FarAndHeight , I'm testing in our sample app by creating a campaign on the dashboard - both by typing straight into the dashboard and by copying the html from an outside text editor and I can't repro. I believe there is an OS X setting that autocorrects into smart apostrophes - this might affect what you're seeing.

**Comment by patrickrills** [comment]
I'm getting black full HTML in-app messages without having the apostrophe in them. Could other characters be causing this as well?

**Comment by FarAndHeight** [comment]
@chshapiro These campaigns have been working for months and as soon as we upgraded to 3.20.1 (we didn't change anything else), they stopped working.  Here's a message that we just tried and it was workingfine, but turned into a blank screen as soon as we introduced the apostrophe after `amet`:
```
<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1 viewport-fit=cover" />
		<title>Modal</title>
		<link href="modal.cssxxx" rel="stylesheet" />
<style>@font-face{font-family:"GT_Medium";src:url("") format("woff");font-style:normal}@font-face{font-family:"GT_bold";src:url("") format("woff");font-style:normal}@font-face{font-family:"GT_Regular";src:url("") format("woff");font-style:normal}*{box-sizing:border-box}html{font-size:62.5%}
@media (min-width: 320px) and (max-width: 374px){html{font-size:56%}
}body{-webkit-font-smoothing:antialiased;font-family:-apple-system,BlinkMacSystemFont,Segoe UI,"Roboto",Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;color:#747474;margin:0;padding:0;font-size:1.6rem}body.default-modal{background-color:rgba(0,0,0,0.4);background-size:cover;background-repeat:no-repeat;height:100vh;display:flex;align-content:center}.container{width:calc(100vw - 32px);height:auto;margin:auto;padding-bottom:24px;border-radius:8px;z-index:1;position:relative;background:#fafafa;overflow:hidden}.button-close{width:24px;height:24px;position:absolute;right:15px;top:15px;z-index:99;background-color:#e5e5e5;border-radius:50%;justify-content:center;display:flex}.button-close img{filter:invert(1);opacity:0.8;width:14px}.btn{align-items:center;background-color:#f5a623;border-radius:24px;color:#fafafa;display:flex;font-family:GT_bold;height:48px;justify-content:center;margin:24px auto 0;max-width:340px;text-align:center;text-decoration:none;width:100%}
h1,h2,h3{font-family:GT_bold,-apple-system,BlinkMacSystemFont,Segoe UI,"Roboto",Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;color:#272727;margin:0}h1{font-size:4rem;margin-bottom:0.8rem}h2{font-size:2.4rem;line-height:3rem;margin:0 18px 0.8rem 0}h3{font-size:2rem}p{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,"Roboto",Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;margin:0 0 1rem 0;padding:0;line-height:2.2rem;color:#595959}.buttons.horizontal{display:flex;align-items:center;justify-content:space-between;margin:24px 0 0}.buttons.horizontal .btn{order:2;height:32px;margin:0;max-width:170px}.buttons.horizontal .actionlink{margin:0;color:#272727}.actionlink{font-size:1.4rem;font-weight:bold;color:#272727;text-align:center;margin:2rem 0 0;text-decoration:none;display:block}.hero{height:calc(50vw - 30px);width:100%;max-height:80vw;overflow:hidden}.hero img{display:block;width:100%;height:100%;object-fit:cover}.copy{padding:24px 24px 0}.container-slide-wrap{top:0;position:relative;transition:ease-in-out top 0.4s}.disclosure{background:#fafafa;height:100px;position:relative}.disclosure:after{content:"";background:linear-gradient(to bottom, rgba(255,255,255,0) 0%, #fafafa 100%);position:absolute;bottom:0;width:100%;height:30px}.disclosure_scroll{height:100%;overflow-y:scroll;-webkit-overflow-scrolling:touch}.disclosure_scroll p{font-size:1.2rem;line-height:1.8rem;margin:0 0 0.8rem}.disclosure_close{justify-content:center;display:flex;height:24px;width:24px;position:absolute;top:15px;left:15px}.disclosure_close img{width:14px}.disclosure.slideup{height:100%;z-index:50;padding:60px 24px 24px;transition:ease-in-out top 0.4s;position:absolute;top:1000px}.disclosure.slideup:after{left:0;bottom:24px;height:40px}

/*# sourceMappingURL=modal.css.map */</style>
	</head>

	<body class="default-modal">
		<div class="container">
			<a href="appboy://close" class="button-close">
				<img src="close.svg" alt="Close" />
			</a>

			<!-- optional section for hero image -->
			<div class="hero">
				<img src="http://placehold.it/375x300&text=2:1" />
			</div>
			<!-- // optional section for hero image -->

			<div class="copy">
				<h2>Lorem ipsum dolor sit amet’, consectetur adipiscing elit.</h2>
				<p>Cras hendrerit ac enim et ullamcorper. Nullam ac ligula enim. Aliquam in diam fermentum, venenatis.</p>

				<!-- remove class 'horizontal' for stacked buttons. remove actionlink if unneeded -->
				<div class="buttons horizontal">
					<a href="#" class="btn small">Action</a>
					<a href="appboy://close" class="actionlink">Maybe later</a>
				</div>
			</div>
		</div>
	</body>
</html>
```

**Comment by TerekhovAnton** [comment]
we experience the same issue, but there are no " ’ " characters in our html.
but it's on iOS 12.4.1.
(just to be sure - sdk version is 3.20.1)

**Comment by chshapiro** [comment]
@TerekhovAnton ,

Are you unable to display your in-app message or unable to interact with it?

**Comment by TerekhovAnton** [comment]
@chshapiro it's a fully white screen shown, no content and unable to interact

**Comment by patrickrills** [comment]
I've tracked down that are issue is including any attachments. If there is a zip with the message it will so blank.

**Comment by chshapiro** [comment]
@TerekhovAnton and @patrickrills , 

We are unable to repro displaying a blank message when including a zip file. Can you email your html and assets zip to jared@braze.com and chava.friedler@braze.com and we will see if we can repro with your specific html?

Thank you!

**Comment by patrickrills** [comment]
It appears that the issue is that the attached files are not getting written to disk cache.

**Comment by chanhpham9291** [comment]
Hi @chshapiro, My project (Unity/iOS) do have some customization on AppboyUnityManager.mm
`- (ABKInAppMessageDisplayChoice)beforeInAppMessageDisplayed:(ABKInAppMessage *)inAppMessage {
	NSLog(@"beforeInAppMessageDisplayed");
	if (self.nativeInAppMessagingEnabled)
	{
		NSLog(@"Displaying Meassage Now");
		return ABKDisplayInAppMessageNow;
	}
	NSLog(@"Display Meassage Later");
    return ABKDisplayInAppMessageLater;
}`

Our blank IAMs work normally if removing `’`.

**Comment by Eirikr8** [comment]
We're getting the same blank view with our inApp messages and it's not a matter of the " ’ ", it's happening since we upgraded to 3.20.1 to the campaigns that were working fine before.

**Comment by patrickrills** [comment]
@Eirikr8 Do you have assets attached to your messages? 

![Screen Shot 2019-09-18 at 9 21 57 AM](https://user-images.githubusercontent.com/7559993/65152344-c82b3300-d9f5-11e9-81f5-4a66c900d6d8.png)

**Comment by Eirikr8** [comment]
Yes, we do have an asset zip attached

**Comment by patrickrills** [comment]
@Eirikr8 That was my issue too. The problem is that the HTML is not getting written to cache with the attachments so the WKWebView is showing empty HTML. I put a temporary fix in our app to ensure its written to cache.

**Comment by juliensechaud** [comment]
@patrickrills how did you make this fix ?

**Comment by patrickrills** [comment]
In ABKInAppMessageHTMLViewController.m:

```objectivec
if (((ABKInAppMessageHTML *)self.inAppMessage).assetsLocalDirectoryPath != nil) {
    NSString *localPath = [((ABKInAppMessageHTML *)self.inAppMessage).assetsLocalDirectoryPath absoluteString];
    // Here we must use fileURLWithPath: to add the "file://" scheme, otherwise the webView won't recognize the
    // base URL and won't load the zip file resources.
    NSURL *html = [NSURL fileURLWithPath:[localPath stringByAppendingPathComponent:ABKInAppMessageHTMLFileName]];

      //TEMP FIX - check for cached HTML; if it doesn't, exist create it
      if (![[NSFileManager defaultManager] fileExistsAtPath:html.absoluteString]) {
          NSData *htmlData = [self.inAppMessage.message dataUsingEncoding:NSUTF16StringEncoding];
          if (htmlData && htmlData.length > 0) {
              NSError *err = nil;
              if (![htmlData writeToURL:html options:NSDataWritingAtomic error:&err]) {
                  NSLog(@"Error writing Braze IAM HTML to cache: %@", (err ? err.description : @"N/A"));
              }
          }
      }

    [self.webView loadFileURL:html allowingReadAccessToURL:[NSURL fileURLWithPath:localPath]];
  } else {
    [self.webView loadHTMLString:self.inAppMessage.message baseURL:nil];
  }
```

**Comment by chshapiro** [comment]
Hey all,

We are planning to release an official solution today. We strongly suggest that you don't implement any workarounds or temporary fixes and instead wait for our release later today.

Thanks!

**Comment by chshapiro** [comment]
Hey all,

We've released version 3.20.2 of our SDK which fixes the HTML in-app message issues you've all been seeing. If you have any further questions, feel free to open a new issue.

Thank you for your patience!

**Comment by juliensechaud** [comment]
@chshapiro thanks for the release. it seems to fix all of it
